File: Meta/MinimumLengthSubstrings/MinimumLengthSubstrings.py
"""
Minimum Length Substrings
You are given two strings s and t. You can select any substring of string s and rearrange the characters of the selected substring. Determine the minimum length of the substring of s such that string t is a substring of the selected substring.
Signature
int minLengthSubstring(String s, String t)
Input
s and t are non-empty strings that contain less than 1,000,000 characters each
Output
Return the minimum length of the substring of s. If it is not possible, return -1
Example
s = "dcbefebce"
t = "fd"
output = 5
Explanation:
Substring "dcbef" can be rearranged to "cfdeb", "cefdb", and so on. String t is a substring of "cfdeb". Thus, the minimum length required is 5.

*** NOTES ***
For substrings/subarrays that don't involve summations, you probably want to use a sliding window. You want to performing an O(m) operation on each window where m is the size of the window however.
This substring problem requires counting some permutation of the substring so counting characters using a dictionary is very valuable
"""

def min_length_substring(s, t):
  # Write your code here
  # Convert input strings to lists of characters
    s = list(s)
    t = list(t)
    min_window = len(t)  # Minimum possible window length is the length of t
    n = len(s)           # Length of the input string

    # Helper function to generate a frequency dictionary for a given character list
    def freq_dict(str_list):
        result = {}
        for char in str_list:
            if char in result:
                result[char] += 1
            else:
                result[char] = 1
        return result

    # Edge case: If substring t is longer than s, return -1 (impossible to find such a window)
    if len(t) > n:
        return -1

    start = 0                             # Start pointer for the sliding window
    t_dict = freq_dict(t)                 # Frequency dictionary of characters in t
    required_count = len(t_dict)          # Number of unique characters in t
    window_count = 0                      # Number of characters in window matching required frequency
    window_dict = {}                      # Frequency dictionary for the current sliding window
    min_length = float('inf')             # Initialize minimum length to infinity

    # Slide the window through s until the end of the string
    for end in range(n):
        end_char = s[end]

        # If end_char is in t_dict, increment its count in window_dict
        if end_char in t_dict:
            if end_char in window_dict:
                window_dict[end_char] += 1
            else:
                window_dict[end_char] = 1

            # If count of end_char in window_dict matches t_dict, increment window_count
            if window_dict[end_char] == t_dict[end_char]:
                window_count += 1

        # When window_count matches required_count, the current window is valid
        while window_count == required_count:
            # Update minimum length with the current valid window size
            min_length = min(min_length, end - start + 1)

            # Shrink the window from the start to potentially find a smaller valid window
            start_char = s[start]
            if start_char in t_dict:
                # If removing start_char reduces a matching count, decrease window_count
                if window_dict[start_char] == t_dict[start_char]:
                    window_count -= 1
                window_dict[start_char] -= 1  # Reduce start_char count in window_dict

                # Remove the character from window_dict if its count drops to zero
                if window_dict[start_char] == 0:
                    del window_dict[start_char]

            start += 1  # Move start pointer forward to shrink the window

    # Return the minimum length found, or -1 if no valid window was found
    return min_length if min_length != float('inf') else -1

File: Meta/MinimumLengthSubstrings/MinimumLengthSubstrings.swift
/*
 Minimum Length Substrings
 You are given two strings s and t. You can select any substring of string s and rearrange the characters of the selected substring. Determine the minimum length of the substring of s such that string t is a substring of the selected substring.
 Signature
 int minLengthSubstring(String s, String t)
 Input
 s and t are non-empty strings that contain less than 1,000,000 characters each
 Output
 Return the minimum length of the substring of s. If it is not possible, return -1
 Example
 s = "dcbefebce"
 t = "fd"
 output = 5
 Explanation:
 Substring "dcbef" can be rearranged to "cfdeb", "cefdb", and so on. String t is a substring of "cfdeb". Thus, the minimum length required is 5.

 *** NOTES ***
 For substrings/subarrays that don't involve summations, you probably want to use a sliding window. You want to performing an O(m) operation on each window where m is the size of the window however.
 This substring problem requires counting some permutation of the substring so counting characters using a dictionary is very valuable
 */

extension String {
    func minLengthOfRearrangedSubstring(containing substring: String) -> Int {
        let s: [Character] = Array(self)     // Convert self (input string) to an array of characters
        let t: [Character] = Array(substring) // Convert substring to an array of characters
        let minWindow = t.count              // Minimum possible window length is the length of t
        let n = s.count                      // Length of the input string
        
        // Helper function to generate a frequency dictionary for a given character array
        func freqDict(_ str: [Character]) -> [Character: Int] {
            var result: [Character: Int] = [:]
            for char in str {
                result[char, default: 0] += 1
            }
            return result
        }
        
        // Edge case: If substring t is longer than the main string s, return -1 (impossible to find such a window)
        if t.count > n {
            return -1
        }
        
        var start = 0                         // Start pointer for the sliding window
        let tDict = freqDict(t)               // Frequency dictionary of characters in t
        let requiredCount = tDict.count       // Number of unique characters in t
        var windowCount = 0                   // Number of characters in window matching the required frequency in tDict
        var windowDict = [Character: Int]()   // Frequency dictionary for the current sliding window
        var minLength = Int.max               // Initialize minimum length to the maximum possible value
        
        // Check if the windowDict already matches tDict initially (though unlikely here)
        if windowDict == tDict {
            return minWindow
        }
        
        // Slide the window through s until the end of the string
        for end in 0..<n {
            let endChar = s[end]
            
            // If endChar is in tDict, increment its count in windowDict
            if let charCount = tDict[endChar] {
                windowDict[endChar, default: 0] += 1
                // If the count of endChar in windowDict matches tDict, increment windowCount
                if windowDict[endChar] == charCount {
                    windowCount += 1
                }
            }
            
            // When windowCount matches requiredCount, the current window is valid
            while windowCount == requiredCount {
                // Update minimum length with the current valid window size
                minLength = min(minLength, end - start + 1)
                
                // Shrink the window from the start to potentially find a smaller valid window
                let startChar = s[start]
                if let charCount = tDict[startChar] {
                    // If removing startChar reduces a matching count, decrease windowCount
                    if windowDict[startChar] == charCount {
                        windowCount -= 1
                    }
                    windowDict[startChar]! -= 1 // Reduce startChar count in windowDict
                }
                start += 1 // Move start pointer forward to shrink the window
            }
        }
        
        // Return the minimum length found, or -1 if no valid window was found
        return minLength == Int.max ? -1 : minLength
    }
}
